src/components/app/index.tsx

```diff
@@ -2,17 +2,58 @@ import { h, Component } from 'preact';
 import { bind, bitmapToImageData } from '../../lib/util';
 import * as style from './style.scss';
 import Output from '../output';
+import Options from '../options';
 
-import {MozJpegEncoder} from '../../lib/codec-wrappers/mozjpeg-enc';
+import { Encoder } from '../../lib/codec-wrappers/codec';
+import { MozJpegEncoder } from '../../lib/codec-wrappers/mozjpeg-enc';
+
+export type ImageType = 'original' | 'jpeg';
+
+export type CodecOptions = any;
+
+type Encoders = {
+  [type: string]: new () => Encoder
+};
+
+const EncoderNames = {
+  original: 'Original Image',
+  jpeg: (options: CodecOptions) => `JPEG ${options.quality || ''}`
+};
+
+const AllEncoders: Encoders = {
+  jpeg: MozJpegEncoder
+};
 
 type Props = {};
 
 type State = {
-  img?: ImageBitmap
+  sourceImg?: ImageBitmap,
+  sourceData?: ImageData,
+  leftImg?: ImageBitmap,
+  rightImg?: ImageBitmap,
+  loading: boolean
+  leftType: ImageType,
+  rightType: ImageType,
+  leftOptions: CodecOptions,
+  rightOptions: CodecOptions
```

Comment:
It's not necessarily bad practice, but it can make it more difficult to implement shouldComponentUpdate intelligently, and people tend to avoid storing objects in state because mutating them in-place can break shouldComponentUpdate entirely (old===new).  I'm not sure we care that much, so I'm open to nesting to make things cleaner.  It would make the methods to update an option simpler, and would also help avoid coupling the whole app to a left/right UI (maybe we have 4 panels down the line).

Reply:
Agreed. In terms of `shouldComponentUpdate`, we could agree to changing the parent if a property needs changing, immutable-style. I think there are libraries to help with that. Do you have a favourite?

Reply:
[immutability-helpers](https://npm.im/immutability-helpers) is decent.